editoast/src/views/documents.rs

```diff
@@ -131,72 +134,91 @@ mod tests {
-            .delete(&mut db_pool.get().await.unwrap())
-            .await
-            .unwrap());
-
-        // Should fail
-        let request = TestRequest::get().uri(&url).to_request();
-        let response = call_service(&service, request).await;
-        assert!(response.status().is_client_error());
-    }
+    use crate::views::tests::create_test_connection_pool;
+    use crate::views::tests::create_test_service_with_connection_pool;
 
     #[derive(Deserialize, Clone, Debug)]
     struct PostDocumentResponse {
         document_key: i64,
     }
 
+    /// Create a document in the database using the connection pool in test mode
+    async fn create_document(connection_pool: &DbConnectionPoolV2) -> Document {
+        let mut conn = connection_pool
+            .get()
+            .await
+            .expect("Failed to get connection pool");
+        Document::changeset()
+            .data("Document post test data".as_bytes().to_vec())
+            .content_type(String::from("text/plain"))
+            .create(&mut conn)
+            .await
+            .expect("Failed to create document")
+    }
+
     #[rstest]
-    async fn document_post(db_pool: Data<DbConnectionPool>) {
-        let service = create_test_service().await;
+    async fn document_post() {
+        let pool = create_test_connection_pool().await;
+        let service = create_test_service_with_connection_pool(pool.clone()).await;
 
-        // Insert document
         let request = TestRequest::post()
             .uri("/documents")
             .insert_header(ContentType::plaintext())
-            .set_payload("Test data".as_bytes().to_vec())
+            .set_payload("Document post test data".as_bytes().to_vec())
             .to_request();
-        let response: PostDocumentResponse = call_and_read_body_json(&service, request).await;
-
-        // Delete the document
-        assert!(
-            Document::delete_static(&mut db_pool.get().await.unwrap(), response.document_key)
-                .await
-                .unwrap()
-        );
+
+        // Insert document
+        let create_response: PostDocumentResponse =
+            call_and_read_body_json(&service, request).await;
+
+        assert!(create_response.document_key > 0);
     }
 
     #[rstest]
-    async fn document_delete(#[future] document_example: TestFixture<Document>) {
-        let document_example = document_example.await;
-        let service = create_test_service().await;
+    async fn get_document() {
+        // Create test connection pool
+        let pool = create_test_connection_pool().await;
+        // Create test service with connection pool
+        let service = create_test_service_with_connection_pool(pool.clone()).await;
+
+        // In test mode, the connection pool will not commit any changes to the database.
+        // Therefore, it is necessary to create all required data before testing the API service.
+        let document = create_document(&pool).await;
```

Comment:
When the test setup is small, I'd rather having it inlined in the test. I get why we factorize the infra setup, or train schedules, etc.. but documents are small enough to not suffer from redundancy. Nothing blocking though :)

Reply:
I tried to inline the creation of the document, but this caused a deadlock. In the function create_document, we get the connection lock guard, and at the end of the function, the guard gets released. However, if we inline the test, it holds the guard, and when we try to call the API, it will also try to get the guard and become blocked.

Reply:
To release the lock we can also do, but maybe a dedicated function is better ?
```rust
let document = {
    let mut conn = pool.get().await.expect("Failed to get connection pool");
    Document::changeset()
        .data(b"Document post test data".to_vec())
        .content_type(String::from("text/plain"))
        .create(&mut conn)
        .await
        .expect("Failed to create document")
};
```

Reply:
Personally I don't mind the scoping block, as it's a reminder that the connection should be dropped as soon as it's not used anymore.

Alternatively this should work as well:

```rust
let document = Document::changeset()
        .data(b"Document post test data".to_vec())
        .content_type(String::from("text/plain"))
        .create(pool.get_ok().deref_mut())
        .await
        .expect("Failed to create document")
```

Maybe @woshilapin can weigh on this?

Reply:
Either way is fine with me. The scoping is usually a sign of a guard (but maybe it's only me who sees this as a sign?). Maybe @leovalais solution is more straightforward for those less familiar with the concept of guards?